plugin/trino-password-authenticators/src/main/java/io/trino/plugin/password/ldap/LdapGroupProvider.java

```diff
@@ -0,0 +1,198 @@
+import java.util.List;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.Set;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+import java.util.stream.Collectors;
+
+import static com.google.common.collect.Iterables.getOnlyElement;
+import static io.trino.collect.cache.SafeCaches.buildNonEvictableCacheWithWeakInvalidateAll;
+import static io.trino.plugin.password.ldap.LdapAuthenticator.containsSpecialCharacters;
+import static io.trino.plugin.password.ldap.LdapAuthenticator.replaceUser;
+import static java.lang.String.format;
+import static java.util.Objects.requireNonNull;
+import static java.util.concurrent.TimeUnit.MILLISECONDS;
+
+public class LdapGroupProvider
+        implements GroupProvider
+{
+    private static final Logger log = Logger.get(LdapGroupProvider.class);
+    private static final Pattern EXTRACT_GROUP_PATTERN = Pattern.compile("(?i)cn=([^,]*),(.*)$");
+
+    private final LdapClient client;
+
+    private final List<String> userBindSearchPatterns;
+    private final Optional<String> userBaseDistinguishedName;
+    private final String groupBaseDistinguishedName;
+    private String groupSearchFilter;
+    private final String bindDistinguishedName;
+    private final String bindPassword;
+    private final boolean allowUserNotExist;
+
+    private final LoadingCache<String, Set<String>> groupCache;
+
+    @Inject
+    public LdapGroupProvider(LdapClient client, LdapGroupProviderConfig ldapConfig)
+    {
+        this.client = requireNonNull(client, "client is null");
+        requireNonNull(ldapConfig, "ldapConfig is null");
+
+        this.userBindSearchPatterns = ldapConfig.getUserBindSearchPatterns();
+        this.userBaseDistinguishedName = Optional.ofNullable(ldapConfig.getUserBaseDistinguishedName());
+        this.groupBaseDistinguishedName = ldapConfig.getGroupBaseDistinguishedName();
+        this.bindDistinguishedName = ldapConfig.getBindDistinguishedName();
+        this.bindPassword = ldapConfig.getBindPassword();
+        this.allowUserNotExist = ldapConfig.isAllowUserNotExist();
+        this.groupSearchFilter = ldapConfig.getGroupSearchFilter();
+        this.groupCache = buildNonEvictableCacheWithWeakInvalidateAll(
+                CacheBuilder.newBuilder()
+                        .expireAfterWrite(ldapConfig.getLdapCacheTtl().toMillis(), MILLISECONDS),
+                CacheLoader.from(this::getGroupsWithBindDistinguishedName));
+    }
+
+    @VisibleForTesting
+    void invalidateCache()
+    {
+        groupCache.invalidateAll();
+    }
+
+    @Override
+    public Set<String> getGroups(String user)
+    {
+        try (ThreadContextClassLoader ignored = new ThreadContextClassLoader(getClass().getClassLoader())) {
+            return groupCache.getUnchecked(user);
+        }
+        catch (UncheckedExecutionException e) {
+            throw e;
+        }
+    }
+
+    private Set<String> getGroupsWithBindDistinguishedName(String user)
```

Comment:
How about AD approach i.e if a user has `memberOf` attribute as a part of its `LdapObject`.

Reply:
OK .. AFAIK there are 3 types  of group membership: 

RFC 2307 (posix) - from 1998, venerable :) - in this draft PR so far
RFC2307bis - Active Directory - rfc never made it out of draft, despite Microsoft using it
Augmented Active Directory.

let me investigate this, see if we can easily add in AD support ...

Reply:
this is a good representation

![image](https://github.com/trinodb/trino/assets/712608/830febed-1c20-4b01-a033-fc7ddb783647)